### PR TITLE
Adding Invert option.

### PR DESCRIPTION
**Pull request by NikolayRys**
Hi! I've needed a feature for my project to make element sticking by lib to the bottom of the screen. So here I am sharing it, please feel free to merge. 
Now it makes available an option "invert", which moves the element to bottom of the viewport and reinterprets option "top", "mind". When this feature is on, they will work as opposite modifying the offset between bottom(not top) of the viewport and element. Also, "mindBottomPadding" will now account top padding instead. This what meant by "invert" semantics. Works only if native sticky is disabled.

Besides that I've did some refactoring, certain performance improvements(less overhead with redundant calculations), couple of small bug fixes, and also updated dependencies, docs and tests.
I've tested it in Chrome and Firefox so far. I guess this all warrants a version bump. For quick demo you could try the dev2.html page, I've added 3 more examples with this feature here.

Please, let me know what do you think.

### CODE REVIEW

File: dist/fixto.js [hunk 1]
```diff
@@ -260,7 +258,7 @@ var fixto = (function ($, window, document) {
     // Checks if browser creates a positioning context for fixed elements.
     // Transform rule will create a positioning context on browsers who follow the spec.
     // Ie for example will fix it according to documentElement
-    // TODO: Other css rules also effects. perspective creates at chrome but not in firefox. transform-style preserve3d effects.
```
Comment (bbarakaci) [inline dist/fixto.js]: preserve3d was not a typo, it is a css value for transfom-value attribute but actually should be preserve-3d.

File: dist/fixto.js [hunk 2]
```diff
@@ -329,7 +328,7 @@ var fixto = (function ($, window, document) {
                     }
                 }
             }
-            return top;
+            return top + this.options.top;
```
Comment (bbarakaci) [inline dist/fixto.js]: this line is violating the description of the method: "Returns the total outerHeight of the elements passed to mind option. Will return 0 if none."

Maybe we change the description if we want to add the top option.

File: src/fixto.js [hunk 3]
```diff
@@ -257,7 +257,7 @@ var fixto = (function ($, window, document) {
     // Checks if browser creates a positioning context for fixed elements.
     // Transform rule will create a positioning context on browsers who follow the spec.
     // Ie for example will fix it according to documentElement
-    // TODO: Other css rules also effects. perspective creates at chrome but not in firefox. transform-style preserve3d effects.
+    // TODO: Other css rules also have effect. perspective creates at chrome but not in firefox. transform-style preserved effects.
```
Comment (bbarakaci) [inline src/fixto.js]: preserve3d was not a typo, it is a css value for transfom-value attribute but actually should be preserve-3d.

File: src/fixto.js [hunk 4]
```diff
@@ -326,7 +327,7 @@ var fixto = (function ($, window, document) {
                     }
                 }
             }
-            return top;
+            return top + this.options.top;
```
Comment (bbarakaci) [inline src/fixto.js:332]: this line is violating the description of the method: "Returns the total outerHeight of the elements passed to mind option. Will return 0 if none."

We will need to change the description if we need to have this.

File: src/fixto.js [hunk 5]
```diff
@@ -396,6 +398,88 @@ var fixto = (function ($, window, document) {
         }
     };
 
+
+    function DirectCalc (container) {
```
Comment (bbarakaci) [inline src/fixto.js:405]: these classes are reaching the internals of the FixToContainer class.

File: src/fixto.js [hunk 6]
```diff
@@ -396,6 +398,88 @@ var fixto = (function ($, window, document) {
...
+            if (this.c.options.mindBottomPadding !== false) { limiter -= computedStyle.getFloat(this.c.parent, 'paddingBottom'); }
+            return limiter;
+        },
+
+        isBetween: function (mindTop) {
+            return this.c._windowLimiter < this.c._parentLimiter &&
+              this.c._windowLimiter > (this.c._fullOffset('offsetTop', this.c.child) - mindTop);
+        },
+
+        initStyle: function (mindTop, margin) {
```
Comment (bbarakaci) [inline src/fixto.js]: i can't see who is using this method.

File: src/fixto.js [hunk 7]
```diff
@@ -396,6 +398,88 @@ var fixto = (function ($, window, document) {
...
+            return limiter;
+        },
+
+        isBetween: function (mindTop) {
+            var isOut = (this.c._windowLimiter - mindTop < this.c._parentLimiter + this.c.child.offsetHeight ) ||
+              (this._viewportTop() + this.c.child.offsetHeight> this.c._fullOffset('offsetTop', this.c.child) + this.c.parent.offsetHeight + mindTop);
+            return !isOut;
+        },
+
+        initStyle: function (mindTop, margin) {
```
Comment (bbarakaci) [inline src/fixto.js]: same above

**Comment by bbarakaci** [comment]
Hello,

I was not planning to add this feature to keep things simple, but now it is in, i need to consider for the sake of the effort you put in.

Many thanks for cleaning up and updating the codebase. grunt was ancient.

Please let me dive into the details when i find some free time, like a weekend, to pull the branch, understand the new logic. There are many changes regarding to calculations.

For the first look, i was thinking, if instead of invert, introducing a bottom option maybe could fallback to native sticky implementation without much hassle.

File: src/fixto.js (hunk 3 above)
Comment (NikolayRys) [reply src/fixto.js]: sure, I'll change it back

File: src/fixto.js (hunk 4 above)
Comment (NikolayRys) [reply src/fixto.js:332]: Yeah, all 100 percent of usages of this method were along with this option. I'll fix the description then

File: src/fixto.js (hunk 5 above)
Comment (NikolayRys) [reply src/fixto.js:405]: I'll give it a second thought, but these classes are actually can be considered as a part of FixToContainer by design. Probably I can achieve better separation of responsibility by pulling out the third class with the purpose of data holding that FixToContainer will create and make DirectCalc and InvertedCalc operate on that.

File: src/fixto.js (hunk 6 above)
Comment (NikolayRys) [reply src/fixto.js]: Yeah, sorry, forgot it from some intermediary code version. I'll remove it.

**Comment by NikolayRys** [comment]
Okay, gotcha. I've implemented not-native version because in my project it was crucial for me to be able to query if the object is fixed or not. If you want me to make it work through native I can do it as well, along with other changes to PR. But two remarks here. 
1) Your initial not-native implementation was a bit slow, because of multiple recalculations with the same values, now this problem is gone which makes native implementation less advantageous in comparison with non-native that it was before. (Btw, the performance can be improved even further, we can discuss it, if you're interested).
2) There's open question with naming, since all your options' names imply that the element is attached to top. So when we call it a "bottom" feature, we need change names all other params. Which "invert" allowed me not to.

File: src/fixto.js (hunk 7 above)
Comment (NikolayRys) [reply src/fixto.js]: ok

**Comment by bbarakaci** [comment]
Hi,

Favoring native implementation over fixto regarding to the performance was for not bothering the mobile devices with scroll events. They were even disabled in ios, and now i am not sure of the state. Supporting the native sticky provided the opportunity to have the feature on all devices through one interface. And the fallback mechanism would let us leave fixto behind once position sticky support is fully there.

You have points over there and it is absolutely logical to think in that manner. It is even possible to support native sticky through the api you have provided. Once it is inverted we can just set the bottom property in native css and it is done.

In case of choosing the bottom option, the only option that bothers is the mindBottomPadding. That was an option provided by another contributor. It could be renamed to ignorePadding and actually it could even be dropped in the future, since such an effect could be achieved by seeting up the html structure and layout accordingly, and it is against native implementation (although could be achieved)

I am interested to hear about the performance optimization but can wait. Since i do not have time because of daily job and family matters, my first priority will be taking your code in.

Best

**Comment by bbarakaci** [comment]
Another idea came to my mind. If you are interested to be a maintainer, i could share my access rights with you so you can merge and release whatever you want. I do not need to be too picky as long as i will not be the sole responsible who will maintain the shipped code.

**Comment by NikolayRys** [comment]
Sure, why not, I would be glad to help you with maintaining. To make a small recap, here's what I'm thinking to do and discuss with you then.
1. Make changes according to your PR comments and test it in IE9. 
2. Extend the same feature to native implementation. That gonna be easy, I guess.
3. Consider(I'm not yet sure) simplifying non-native implementation using http://imakewebthings.com/waypoints/shortcuts/sticky-elements/
Since support for native sticky is not expected to become ubiquitous for very long time, I think this might be a worthwhile effort that will greatly ease the maintaining for us.
4. Publish it also as an npm package for convenience.
5. The version now is 0.5, but the plugin exists for a long time. What's left to make it release version 1.0.0?
6. Analyse the performance issue I've mentioned above. I've noticed that in some browsers (at least desktop chrome and firefox, but probably not only) the "onscroll" event is triggered much more than necessary with the same params, adding presumably heavy recalculation burden. We could filter out those excessive calls - I have a couple of ideas how. Probably just save the old calculated values and return them if the current viewport state is the same.
7. And I totally agree with you in dropping mindBottomPadding, it indeed feels awkward and out of responsibility of a plugin for stickiness.

**Comment by NikolayRys** [comment]
Did several cosmetical changes as you have requested. The only thing, I haven't yet changed that Calc classes reach for internals of FixToContainer. It's a legitimate strategy pattern where I've passed all data needed for calculation as one data structure, that happened to be FixToContainer. If you still would like to change it, let's find some time to discuss it - there are several options how it can be done, but I personally like them a bit less then the current approach. As an example of what I'm talking about is for example to bind the methods in Calc classes to FixToContainer to make "this" in them refer to it. Or pass all data as parameters every time.

**Comment by NikolayRys** [comment]
Cleaned it up as much as possible.

**Comment by bbarakaci** [comment]
Hi,

3. Why not, as long as the plugin offers the functionality, it is ok to use anything behind the scenes, but, 1, the external library should support amd or commonjs or export because i do not want to introduce any other thing to the browsers window object, 2, it must be jQuery independent because i want to make it usable without jQuery, while still supporting jQuery. I have a local branch eliminates all jQuery usage.

4. I agree.

5. Simply, i do not find it as a mature software good enough for production. It is my way of telling people that they should use it their own risk. A better defined api and test coverage might improve it.

6. I did not witness browsers triggering the scroll event with same values. Maybe you observed it on a page that has multiple instances. As every instance sets its own event listener, then you will observe multiple listeners responding to that. It could be resolved by implementing one listener service that all the instances will benefit. Still, each instance has to its own calculations, so you will save only multiple scrollTop checks.

I have a local branch introducing some infrastructure improvements. It uses es6 through babel, modules are separated to their own files, qunit and jasmine tests are dropped, implemented selenium tests, using npm scripts instead of grunt, eliminates jQuery. This branch is sitting there for a long time because i do not find time to work on this.

**Comment by bbarakaci** [comment]
Oh i forgot to mention another very important reason that i am not sure to further improve this plugin.

Modern javascript stacks are really going fast, starting with angular, nowadays react and many others. These frameworks/libraries usually prefer to do their own magic, instead of letting the developer manipulating/querying the dom manually. For example, to be able to use this plugin in said frameworks you need to write an angular directive or react component around it. Those communities already went with their own solutions and created these components, and on the traditional side now i learned from you that Waypoints has a similar feature. So i doubt the community still needs fixto. Do they?

**Comment by NikolayRys** [comment]
Good question, but here's how I see it now: waypoint.sticky is a quite simple lib that does not work as a satisfactory polyfill for css position:sticky. It allows only to attach it to the top of the viewport and also pulls jquery with it. It does not have "mind" option which is also very useful and does not acknowledge the native implementation that should be used always when available. So I agree that it still reduces to a certain extent the demand for fixto, but it still retains its own niche and if we, after simplifying code with basic waypoints(not waypoints.sticky), add left and right options, it will be viable, lightweight and useful for at least couple years, I think. Until ie dies and edge with mobile browsers will support the native.
The regular waypoints are good for detection if the element is visible so it'll allow us to remove all logic responsible for that.

**Comment by bbarakaci** [comment]
I had the chance to pull the branch and check the dev2 page.

This implementation maybe works for your needs but the result is not compatible with a sticky positioned to bottom. If you view with firefox, you can easily place a native position sticky container with bottom: 0 and observe the behaviour.

In this implementation the placeholder which is created by the ghost node is visible. A part of the content is at all times inaccesible. Notice that in the normal calculation we never see that space. It is there to prevent the jumping of the content when sticky goes out of the flow.

User usually will place the container to the bottom of the content for invertd usage. In that case everything goes well, but the adjusting part breaks, container disappears. You can observe this by placing the sticy after the lorem ipsum.

You can also put it to the middle of the content to see the behaviour.

Actually the inverted version makes it fixed at all times, even when all the parent container is in the viewport and no edge of it touches to the viewport.

If this behaviour is what you have needed, i am glad you have it now but i am afraid i can not take it in.

File: dist/fixto.js [hunk 8]
```diff
@@ -1,7 +1,5 @@
-/*! fixto - v0.5.0 - 2016-06-16
```
Comment (bbarakaci) [inline dist/fixto.js:1]: if you happen to push more commits, please do not add this file until releasing. i usually have a release commit. Until then it should not be in the history. It is also making the pr confusing.

The version number also will be incremented while preparing this file. Currently all these are maintained manually.

**Comment by NikolayRys** [comment]
Ah, I see the issue. Didn't need it myself, but that's easily fixable. I will have time for that tomorrow.

File: dist/fixto.js (hunk 8 above)
Comment (NikolayRys) [reply dist/fixto.js:1]: okay

**Comment by NikolayRys** [comment]
sorry, had loads of work for the last 2 weeks, gonna handle this fix asap

**Comment by bbarakaci** [comment]
Hi. No worries.

I suggest you not to invest more time on this at the moment. I had the chance to think about it. I think i should favor first the software, then myself and then other people's efforts.

I don't think i can afford the maintenance cost of the changes in the code. I thought we could resolve it by adding you as collaborator but this pr showed me that it can only work in the long run. We had to spend significant time on this pr and it is not even close to get in on many ways.

I am planning to push this branch of yours to the main repository so your commits will stay. Please let me know if you object. I will think about the rest in the future, after i introduce the changes i have mentioned earlier.

Sorry about this. Do not hesitate to contact me personally.
Best

**Comment by NikolayRys** [comment]
Okay, that's a bit sad, however. I still would have preferred to finish it, since we've started and already went so far. But okay, if you don't see it to be fit right now and want to introduce your changes first, then please ping me when they are done and you're ready to finish it. Of course, if native approach is not yet there.
For now, maybe, I'll create a second PR for you that contain only initial commits with grunt and dependencies update?

**Comment by NikolayRys** [comment]
Please let me know what do you think about this intermediary option.

**Comment by bbarakaci** [comment]
I pushed my local branch with the name dev. Please have a look.

According to that, grunt is leaving it's place to npm scripts and there are many other architectural changes.

Then you decide if said pr is necessary. I will appreciate if you have opinions about the new version.

**Comment by NikolayRys** [comment]
Yes, it does, it can use grunt update:) I'll prepare it this evening.

**Comment by NikolayRys** [comment]
Okay, I've made a micro PR there. So when your code is ready to be extended with the `invert` strategy, if you haven't change your mind, ping me, pls, I'll add it, using approach from this branch and fix the issue with `bottom: 0`.
Thanks.

**Pull request closed without merging**